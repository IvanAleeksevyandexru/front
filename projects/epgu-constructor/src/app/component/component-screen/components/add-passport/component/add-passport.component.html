<epgu-constructor-screen-pad class="screen-pad-body">
  <section class="add-passport-screen">
    <div class="add-passport-screen__body">
      <epgu-constructor-passport
        [attrs]="$any(data)"
      >
      </epgu-constructor-passport>
    </div>
  </section>
</epgu-constructor-screen-pad>
