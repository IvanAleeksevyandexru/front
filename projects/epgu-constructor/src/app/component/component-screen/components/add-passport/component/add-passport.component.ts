import { ChangeDetectionStrategy, Component, Input } from '@angular/core';

import { ComponentBase } from '../../../../../screen/screen.types';

@Component({
  selector: 'epgu-constructor-add-passport',
  templateUrl: './add-passport.component.html',
  styleUrls: ['./add-passport.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class AddPassportComponent {
  @Input() data: ComponentBase;
}
