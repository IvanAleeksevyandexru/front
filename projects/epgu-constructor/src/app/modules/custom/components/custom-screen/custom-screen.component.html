<epgu-constructor-navigation class="nav"></epgu-constructor-navigation>
<epgu-constructor-screen-container class="screen-container">
  <div class="inner">
    <div *ngIf="data?.header" class="head">
      <epgu-constructor-page-name>{{ data?.header }}</epgu-constructor-page-name>
    </div>

    <epgu-constructor-screen-pad class="body">
      <form #customForm="ngForm" action="" class="form">
        <ng-container *ngFor="let componentData of data?.components">
          <ng-container [ngSwitch]="componentData.type">
            <div class="form__item" *ngSwitchCase="componentType.LabelSection">
              <epgu-constructor-label-section
                class="label"
                [name]="componentData.label"
              ></epgu-constructor-label-section>
            </div>

            <div class="form__item" *ngSwitchCase="componentType.Dictionary">
              <div class="form__title">{{ componentData.label }}</div>
              <lib-dropdown
                class="cs__select cs-select"
                (changed)="selectDictionary($event, componentData)"
                [items]="dictionary[componentData?.attrs?.dictionaryType]?.list"
              ></lib-dropdown>
            </div>

            <div class="form__item" *ngSwitchCase="componentType.Lookup">
              <div class="form__title">{{componentData.label}}</div>
              <lib-lookup
                [queryMinSymbolsCount]="0"
                [searchCaseSensitive]="false"
                (changed)="selectDictionary($event, componentData)"
                [fixedItems]="dictionary[componentData?.attrs?.dictionaryType]?.list"></lib-lookup>
            </div>

            <div class="form__item" *ngSwitchCase="componentType.StringInput">
              <div class="form__title">{{ componentData.label }}</div>
              <lib-plain-input
                [invalid]="!state[componentData.id].valid"
                [name]="componentData.id"
                (change)="inputChange($event, componentData)"
                [validationShowOn]="validationShowOn"
              ></lib-plain-input>
              <div class="form__error">
                <small *ngIf="!state[componentData.id].valid"
                  >{{ state[componentData.id]?.errorMessage }}
                </small>
              </div>
            </div>

            <div class="form__item" *ngSwitchCase="componentType.RadioInput">
              <div class="form__title">{{ componentData.label }}</div>
              <div class="form__radio__block">
                <lib-radio *ngFor="let item of componentData?.attrs?.supportedValues"
                  [name]="componentData.id"
                  [labelText]="item.label"
                  (change)="inputChange($event, componentData)"
                  [value]="item.value">

                </lib-radio>
              </div>
            </div>

            <div class="form__item" *ngSwitchCase="componentType.RadioButton">
              <div class="form__title">{{ componentData.label }}</div>
              <div class="form__radio__inline">
                <lib-radio
                  [name]="componentData.id"
                  (change)="inputChange($event, componentData)"
                  [value]="true"
                  labelText="Да">

                </lib-radio>
                <lib-radio
                  [name]="componentData.id"
                  (change)="inputChange($event, componentData)"
                  [value]="false"
                  labelText="Нет">

                </lib-radio>
              </div>
            </div>

            <div class="form__item" *ngSwitchCase="componentType.DateInput">
              <div class="form__title">{{ componentData.label }}</div>
              <lib-date-picker
                [name]="componentData.id"
                [(ngModel)]="state[componentData.id].value"
              ></lib-date-picker>
            </div>

            <div class="form__item" *ngSwitchCase="componentType.RadioInput">
              <div class="form__title">{{ componentData.label }}</div>
              <epgu-constructor-radio-input *ngFor="let radio in componentData.attrs.supportedValues">
                <lib-radio [value]="radio.value" [labelText]="radio.label"></lib-radio>
              </epgu-constructor-radio-input>
            </div>

          </ng-container>
        </ng-container>
      </form>
    </epgu-constructor-screen-pad>

    <div class="footer">
      <lib-button
        [showLoader]="constructorService.isLoading"
        (click)="nextScreen()"
        class="footer__btn-submit"
        >{{ data.submitLabel }}</lib-button
      >
    </div>
  </div>
</epgu-constructor-screen-container>
