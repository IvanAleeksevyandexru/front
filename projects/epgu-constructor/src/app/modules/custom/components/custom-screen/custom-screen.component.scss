@import "src/styles/core";


.nav {
  @include navigation-mt();
}

.inner {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  height: inherit;
  min-height: inherit;
  @include tablets-up {
    min-height: auto;
  }
}

.body {
  display: flex;
  text-align: left;
  @include tablets-down {
    flex-grow: 1;
  }
  > * {
    display: flex;
    width: 100%;
    flex-direction: column;
  }
}

.label {
  margin-bottom: 20px;
}

.form {
  &__item {
    margin-bottom: 24px;
  }
  &__title {
    font-size: 14px;
    line-height: 20px;
    color: $color-summer-night;
    margin-bottom: 4px;
  }
  &__error {
    margin-top: 4px;
    color: $color-gosred;
  }
  &__radio {
    &__block {
      display: inline-block;
      lib-radio {
        display: block;
        margin-bottom: 16px;
      }
    }
    &__inline {
      margin-top: 24px;
      display: inline-flex;
      lib-radio:not(:first-child) {
        margin-left: 24px;
      }
    }
  }
}

.footer {
  display: flex;
  align-items: center;
  justify-content: center;
  margin-top: 24px;
  @include tablets-up {
    margin-top: 32px;
  }
  &__btn-submit {
    width: auto;
    &::ng-deep button {
      width: inherit;
      min-width: 100px;
    }
  }
}
