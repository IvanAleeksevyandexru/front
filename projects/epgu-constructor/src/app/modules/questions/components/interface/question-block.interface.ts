import {
  EgpuResponseComponentInterface,
  EgpuResponseDisplayInterface,
  EgpuResponseInterface
} from '../../../../../interfaces/epgu.service.interface';


export interface EgpuResponseQuestionsDisplayInterface extends EgpuResponseDisplayInterface {
  components: Array<EgpuResponseQuestionsDisplayComponentInterface>;
}

interface EgpuResponseQuestionsDisplayComponentInterface extends EgpuResponseComponentInterface {
  attrs: {
    actions: Array<EgpuResponseQuestionsDisplayComponentAttrsActionsInterface>
  }
}

export interface EgpuResponseQuestionsDisplayComponentAttrsActionsInterface {
  action: string;
  label: string;
  value: string;
}
