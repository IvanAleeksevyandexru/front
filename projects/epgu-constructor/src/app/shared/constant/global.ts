export enum SCREEN_TYPE {
  'QUESTION' = 'QUESTION',
  'INFO' = 'INFO',
  'COMPONENT' = 'COMPONENT', // внутри этого компонента внутри ровна один компонент
  'CUSTOM' = 'CUSTOM',
  'UNIQUE' = 'UNIQUE',
  'INVITATION_ERROR' = 'INVITATION_ERROR',
  'EMPTY' = 'EMPTY'
}

export enum SCREEN_COMPONENT_NAME {
  'invitationError' = 'InvitationError',
  'confirmPersonalUserRegAddr' = 'ConfirmPersonalUserRegAddr',
  'confirmPersonalUserData' = 'ConfirmPersonalUserData',
  'confirmPersonalUserEmail' = 'ConfirmPersonalUserEmail',
  'confirmPersonalUserPhone' = 'ConfirmPersonalUserPhone',
  'registrationAddr' = 'RegistrationAddr',
  'personLookup' = 'PersonLookup',
  'addPassport' = 'AddPassport',
  'changeList' = 'ChangeList',
  'childrenListUnder14' = 'ChildrenListUnder14',
  'childrenListAbove14' = 'ChildrenListAbove14',
  'snilsInput' = 'SnilsInput',
  'docInput' = 'DocInput',
  'countryApostil' = 'CountryApostil',
  'paymentScr' = 'PaymentScr'
}

export enum INFO_SCREEN_COMPONENT {
  'info' = 'InfoScr',
}

export enum UNIQUE_COMPONENT_NAME {
  fileUploadComponent = 'FileUploadComponent',
  'addPassport' = 'AddPassport',
  'timeSlot' = 'TimeSlot',
  'mapService' = 'MapService',
  'carInfo' = 'CarInfo',
  'repeatableFields' = 'RepeatableFields',
  'employeeHistory' = 'EmployeeHistory',
  'confirmMarriage' = 'ConfirmMarriage',
}

export enum EMPTY_SCREEN_COMPONENT {
  'redirect' = 'Redirect',
}

export enum CUSTOM_COMPONENT_ITEM_TYPE {
  LabelSection = 'LabelSection',
  Dictionary = 'Dictionary',
  DropDown = 'DropDown',
  StringInput = 'StringInput',
  DateInput = 'DateInput',
  RadioInput = 'RadioInput',
  CompositeRadioButton = 'CompositeRadioButton',
  Lookup = 'Lookup',
  AddressInput = 'AddressInput',
}

export const DATE_STRING_DOT_FORMAT = 'DD.MM.YYYY';
export const DATE_STRING_DASH_FORMAT = 'DD-MM-YYYY';
