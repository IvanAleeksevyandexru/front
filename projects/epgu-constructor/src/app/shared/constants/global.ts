export enum SCREEN_TYPE {
  'QUESTION' = 'QUESTION',
  'INFO' = 'INFO',
  'COMPONENT' = 'COMPONENT', // внутри этого компонента внутри ровна один компонент
  'CUSTOM' = 'CUSTOM',
  'UNIQUE' = 'UNIQUE',
  'INVITATION_ERROR' = 'INVITATION_ERROR',
  'EMPTY' = 'EMPTY'
}

export enum SCREEN_COMPONENT_NAME {
  'invitationError' = 'InvitationError',
  'confirmPersonalUserRegAddr' = 'ConfirmPersonalUserRegAddr',
  'confirmPersonalUserData' = 'ConfirmPersonalUserData',
  'confirmPersonalUserEmail' = 'ConfirmPersonalUserEmail',
  'confirmPersonalUserPhone' = 'ConfirmPersonalUserPhone',
  'registrationAddr' = 'RegistrationAddr',
  'personLookup' = 'PersonLookup',
  'addPassport' = 'AddPassport',
  'changeList' = 'ChangeList',
  'childrenListUnder14' = 'ChildrenListUnder14',
  'childrenListAbove14' = 'ChildrenListAbove14',
  'snilsInput' = 'SnilsInput',
  'docInput' = 'DocInput',
  'countryApostil' = 'CountryApostil',
  'paymentScr' = 'PaymentScr'
}

export enum INFO_SCREEN_COMPONENT {
  'info' = 'InfoScr',
}
