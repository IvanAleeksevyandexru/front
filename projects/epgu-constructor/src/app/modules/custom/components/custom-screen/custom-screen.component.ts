import { Component, EventEmitter, Input, OnChanges, Output, SimpleChanges } from '@angular/core';
import { ListItem, ValidationShowOn } from 'epgu-lib';
import * as moment_ from 'moment';
import { takeUntil } from 'rxjs/operators';
import { DATE_STRING_DOT_FORMAT } from '../../../../../constant/global';
import {
  CustomComponentDictionaryState,
  CustomComponentInterface,
  CustomComponentState,
  CustomDisplayInterface,
} from '../../../../../interfaces/custom-component.interface';
import { DictionaryResponse } from '../../../../../interfaces/dictionary-options.interface';
import { NavigationService } from '../../../../shared-module/service/navigation/navigation.service';
import { ConstructorService } from '../../../../services/constructor/constructor.service';
import { RestService } from '../../../../services/rest/rest.service';
import { UnsubscribeService } from '../../../../services/unsubscribe/unsubscribe.service';
import {
  CUSTOM_COMPONENT_ITEM_TYPE,
  getCustomScreenDictionaryFirstState,
  getNormalizeDataCustomScreenDictionary,
} from '../../tools/custom-screen-tools';

const moment = moment_;

@Component({
  selector: 'epgu-constructor-custom-screen',
  templateUrl: './custom-screen.component.html',
  styleUrls: ['./custom-screen.component.scss'],
  providers: [UnsubscribeService],
})
export class CustomScreenComponent implements OnChanges {
  // <-- constant
  componentType = CUSTOM_COMPONENT_ITEM_TYPE;

  // <-- variables
  validationShowOn = ValidationShowOn.TOUCHED_UNFOCUSED;
  state: { [key: string]: CustomComponentState } = {};
  dictionary: { [key: string]: CustomComponentDictionaryState } = {};

  @Input() data: CustomDisplayInterface;
  @Input() errors: object;
  @Output() nextStepEvent = new EventEmitter();
  @Output() prevStepEvent = new EventEmitter();

  constructor(
    private restService: RestService,
    private navService: NavigationService,
    public constructorService: ConstructorService,
    private ngUnsubscribe$: UnsubscribeService,
  ) {
    this.navService.clickToBack$
      .pipe(takeUntil(this.ngUnsubscribe$))
      .subscribe(() => this.prevStep());
  }

  prevStep() {
    this.prevStepEvent.emit();
  }

  ngOnChanges(changes: SimpleChanges): void {
    if (changes?.data?.currentValue) {
      this.state = {};
      this.data.components.forEach((component) => {
        if (component.type !== CUSTOM_COMPONENT_ITEM_TYPE.LabelSection) {
          this.initState(component);
        }
        if (component.type === CUSTOM_COMPONENT_ITEM_TYPE.Dictionary) {
          const dictionaryName = component.attrs.dictionaryType;
          this.initDictionary(dictionaryName);
          this.loadDictionary(dictionaryName, component);
        }
      });
    }
  }

  nextScreen() {
    // TODO добавить валидацию и проверку заполнения всех полей помимо StringInput
    const responseData = {};
    let isValid = true;
    Object.keys(this.state).forEach((key) => {
      if (this.state[key].component.type === CUSTOM_COMPONENT_ITEM_TYPE.StringInput) {
        const inputValidationResult = this.checkInputValidation(
          this.state[key].value,
          this.state[key].component,
        );
        this.setValidationState(
          inputValidationResult,
          this.state[key]?.component?.id,
          this.state[key]?.value,
        );
        if (inputValidationResult > -1) {
          isValid = false;
        }
      } else if (this.state[key].component.type === CUSTOM_COMPONENT_ITEM_TYPE.DateInput) {
        this.state[key].value = moment(this.state[key].value).format(DATE_STRING_DOT_FORMAT);
      }
    });

    this.validationShowOn = ValidationShowOn.IMMEDIATE;

    if (isValid) {
      Object.keys(this.state).forEach((key) => {
        const responseValue = this.state[key].value;
        responseData[key] =
          typeof responseValue === 'object'
            ? { visited: true, value: JSON.stringify(responseValue || {}) }
            : { visited: true, value: responseValue };
      });
      this.nextStepEvent.emit(responseData);
      this.validationShowOn = ValidationShowOn.TOUCHED_UNFOCUSED;
    }
  }

  selectDictionary(selectedItem: ListItem, component: CustomComponentInterface) {
    const dictionaryName = component.attrs?.dictionaryType;
    this.dictionary[dictionaryName].selectedItem = selectedItem.originalItem;
    this.state[component.id].value = selectedItem.originalItem;
    this.state[component.id].valid = true;
  }

  setValidationState(inputValidationResult, componentId, componentValue) {
    const handleSetState = (isValid, errMsg?) => {
      this.state[componentId].value = componentValue;
      this.state[componentId].valid = isValid;
      this.state[componentId].errorMessage = errMsg;
    };

    if (inputValidationResult === -1) {
      handleSetState(true);
    } else {
      handleSetState(
        false,
        this.state[componentId]?.component.attrs.validation[inputValidationResult].errorMsg,
      );
    }
  }

  inputChange($event: Event, component: CustomComponentInterface) {
    const { value } = $event.target as HTMLInputElement;
    const inputValidationResult = this.checkInputValidation(value, component);

    this.setValidationState(inputValidationResult, component.id, value);
  }

  checkInputValidation(value: string, component: CustomComponentInterface): number {
    const regExpArr = component?.attrs?.validation?.map((item) => {
      try {
        return new RegExp(item.value);
      } catch {
        console.error(`Неверный формат RegExp выражения: ${item.value}. Заменено на /.*/`);
        return new RegExp(/.*/);
      }
    });

    let result = -1; // if result === -1 input value is considered valid

    if (regExpArr) {
      regExpArr.every((regExp, index) => {
        if (!regExp.test(value)) {
          result = index;
          return false;
        }
        return true;
      });
    }

    return result;
  }

  // dateChange($event: any, componentData: ComponentInterface) {
  //   console.log($event, componentData)
  // }

  loadDictionary(dictionaryName: string, component: CustomComponentInterface) {
    // TODO добавить обработку loader(-а) для словарей и ошибок;
    this.restService.getDictionary(dictionaryName, { pageNum: 0 }).subscribe(
      (data) => this.loadDictionarySuccess(dictionaryName, data, component),
      () => this.loadDictionaryError(dictionaryName),
    );
  }

  loadDictionarySuccess(
    key: string,
    data: DictionaryResponse,
    component: CustomComponentInterface,
  ) {
    this.dictionary[key].loading = false;
    this.dictionary[key].paginationLoading = false;
    this.dictionary[key].data = data;
    this.dictionary[key].origin = component;
    this.dictionary[key].list = getNormalizeDataCustomScreenDictionary(data.items, key);
  }

  loadDictionaryError(key: string) {
    this.dictionary[key].loading = false;
    this.dictionary[key].paginationLoading = false;
    this.dictionary[key].loadError = true;
    this.dictionary[key].loadEnd = false;
  }

  // <------------ tools
  initDictionary(dictionaryName) {
    this.dictionary[dictionaryName] = getCustomScreenDictionaryFirstState();
  }

  private initState(component: CustomComponentInterface) {
    const { id, value } = component;
    this.state[id] = { valid: false, errorMessage: '', value, component };
  }
}
