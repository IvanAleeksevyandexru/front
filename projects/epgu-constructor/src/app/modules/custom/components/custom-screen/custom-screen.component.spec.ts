import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core'

import { CustomScreenComponent } from './custom-screen.component';
import { RestService } from '../../../../services/rest/rest.service'
import { RestServiceStub } from '../../../../services/rest/rest.service.stub'
import { NavigationService } from '../../../../layout/service/navigation/navigation.service'
import { ConstructorService } from '../../../../services/constructor/constructor.service'
import { ConstructorServiceStub } from '../../../../services/constructor/constructor.service.stub'
import { EgpuResponseCustomComponentDisplayInterface } from '../../../../../interfaces/custom-component.interface'
import { COMPONENT_TYPE } from '../../../../../constant/global'


describe('CustomScreenComponent', () => {
  let component: CustomScreenComponent;
  let fixture: ComponentFixture<CustomScreenComponent>;
  let restService: RestService;
  let navService: NavigationService;
  let constructorService: ConstructorService;
  let mockData: EgpuResponseCustomComponentDisplayInterface = {
    components: [],
    header: '',
    id: '',
    name: '',
    submitLabel: '',
    type: COMPONENT_TYPE.COMPONENT
  };

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      schemas: [ CUSTOM_ELEMENTS_SCHEMA ], // TODO: remove this line when resolve issue with @ifc/plugin and @ifc/common dependencies
      declarations: [ CustomScreenComponent ],
      providers: [
        NavigationService,
        {provide: RestService, useClass: RestServiceStub},
        {provide: ConstructorService, useClass: ConstructorServiceStub},
      ]
    })
    .compileComponents();
    restService = TestBed.inject(RestService);
    navService = TestBed.inject(NavigationService);
    constructorService = TestBed.inject(ConstructorService);
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(CustomScreenComponent);
    component = fixture.componentInstance;
    component.data = mockData;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });
});
