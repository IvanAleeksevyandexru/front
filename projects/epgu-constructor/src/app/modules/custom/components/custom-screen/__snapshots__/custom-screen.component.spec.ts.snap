// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomScreenComponent check snapshot 1`] = `
<epgu-constructor-custom-screen
  __ngContext__={[Function LRootView]}
  constructorService={[Function ConstructorServiceStub]}
  cycledFieldsKeys={[Function Array]}
  data={[Function Object]}
  dataToSend={[Function Object]}
  flattenCycledFieldsValues={[Function Object]}
  isCycledFields="false"
  navService={[Function NavigationService]}
  nextStepEvent={[Function EventEmitter_]}
  ngUnsubscribe$={[Function UnsubscribeService]}
  prevStepEvent={[Function EventEmitter_]}
>
  <epgu-constructor-navigation
    class="nav"
  /><epgu-constructor-screen-container
    class="screen-container"
  >
    <div
      class="container"
    >
      <div
        class="inner"
      >
        
        <epgu-constructor-screen-pad
          class="body"
        >
          <epgu-constructor-components-list />
        </epgu-constructor-screen-pad>
        <div
          class="footer"
        >
          <lib-button
            class="footer__btn-submit"
          >
            
            <div
              class="button-container"
            >
              <button
                class="button font-"
                tabindex="0"
              >
                
                
                
                
                
              </button>
              
              
              
              
            </div>
          </lib-button>
        </div>
      </div>
    </div>
  </epgu-constructor-screen-container>
</epgu-constructor-custom-screen>
`;
